lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```diff
@@ -0,0 +1,338 @@
+
+        in.setPosition(pos);
+        chars.chars[wordStart] = (char) in.readVInt();
+        int prevPos = pos - in.readVInt();
+
+        int dataLength = in.readVInt();
+        if (forms.ints.length < dataLength) {
+          forms.ints = new int[dataLength];
+        }
+        readForms(forms, in, dataLength);
+
+        int afterForms = in.getPosition();
+
+        while (prevPos != 0 && wordStart > 0) {
+          in.setPosition(prevPos);
+          chars.chars[--wordStart] = (char) in.readVInt();
+          prevPos -= in.readVInt();
+        }
+
+        if (wordStart > 0) {
+          chars.offset = wordStart;
+          chars.length = maxLength - wordStart;
+          processor.accept(chars, forms);
+        }
+
+        if (!collision) {
+          break;
+        }
+
+        in.setPosition(afterForms);
+        collision = in.readVInt() == 1;
+        pos -= in.readVInt();
+      }
+    }
+  }
+
+  private boolean isSameString(
+      char[] word, int offset, int length, int dataPos, ByteArrayDataInput in) {
+    for (int i = length - 1; i >= 0; i--) {
+      in.setPosition(dataPos);
+      char c = (char) in.readVInt();
+      if (c != word[i + offset]) {
+        return false;
+      }
+      dataPos -= in.readVInt();
+      if (dataPos == 0) {
+        return i == 0;
+      }
+    }
+    return length == 0;
+  }
+
+  private void readForms(IntsRef forms, ByteArrayDataInput in, int length) {
+    for (int i = 0; i < length; i++) {
+      forms.ints[i] = in.readVInt();
+    }
+    forms.length = length;
+  }
+
+  static class Builder {
+    private final boolean hasCustomMorphData;
+    private final int[] hashTable;
+    private final int[] chainLengths;
+    private final FlagEnumerator flagEnumerator;
+    private final ByteArrayDataOutput dataWriter;
+
+    private byte[] wordData;
+
+    private int commonPrefixLength, commonPrefixPos;
+    private String currentEntry = null;
+    private final List<char[]> group = new ArrayList<>();
+    private final List<Integer> morphDataIDs = new ArrayList<>();
+
+    Builder(int wordCount, boolean hasCustomMorphData, FlagEnumerator flagEnumerator) {
+      this.flagEnumerator = flagEnumerator;
+      this.hasCustomMorphData = hasCustomMorphData;
+
+      hashTable = new int[wordCount];
+      wordData = new byte[wordCount * 6];
+
+      dataWriter = new ByteArrayDataOutput(wordData);
+      dataWriter.writeByte((byte) 0); // zero index is root, contains nothing
+      chainLengths = new int[hashTable.length];
+    }
+
+    void add(String entry, char[] flags, int morphDataID) throws IOException {
+      if (!entry.equals(currentEntry)) {
+        if (currentEntry != null) {
+          if (entry.compareTo(currentEntry) < 0) {
+            throw new IllegalArgumentException("out of order: " + entry + " < " + currentEntry);
+          }
+          int pos = flushGroup();
+
+          commonPrefixLength = GeneratingSuggester.commonPrefix(currentEntry, entry);
```

Comment:
Hmm it's weird this `commonPrefixLength` is a member of this class instead of a local variable, and we could pass it to `flushGroup`?

Edit: sorry, nevermind: there is an off-by-one here, where we need the `commonPrefixLength` of the *last* entry before the current one which is different from the previous "group", OK.